valtech/aem-easy-content-upgrade
# working change listener based cloud migration trigger

As I mentioned in #172 and #171 I have created a way of migrating content using a `ResourceChangeListener`.

I added the code in this PR after changing it a bit to match the AECU way of using a ResourceResolver, etc.
It can serve as a guide on how AECU approaches the content migration on cloud.

This service is tested on AEM 6.5, cloud-sdk AEM but also on AEMaaCS.  We have verified that the migration is triggered correctly on author when deploying and also on publish after automatic replication in the cloud.

All scripts are executed, taking into account their config, execution history, etc.

How it works:

- service listens for a certain node to be added in path `/conf/aecu/trigger`. The name of this node is configurable (default=`migration-trigger`), the type should be `nt:unstructured`.
- when the resourcechangelistener is triggered scripts are executed using the `AecuServiceMBean`, so taking into account their history, config, etc.
- the service, however, also triggers a first time when it is activated (on deploy)
- the trigger-node is automatically deleted after migration
- the service is also a factory so multiple trigger-points can be configured (if needed)

To test this service
1. install the bundle
5. provide an OSGi config for the service (the default value will do)
6. create the path `/conf/aecu/trigger` (sling:Folder) and then create a node `migration-trigger` (nt:unstructured) below it.
7. This will trigger the migration.

To fully test this on cloud AEM so the migration is triggered on-deploy:
1. create the `/conf/aecu/trigger/migration-trigger` path (with correct node types) automatically
* we did this from our `ui.scripts` or `ui.content` module but a RepoInit script should work aswell
2. when the `AecuMigrationTrigger` service activates on author the scripts are executed a first time (so on deploy)
3. this should automatically be replicated to publish
4. the migration can be re-triggered on demand by creating the `/conf/aecu/trigger/migration-trigger` node with the correct node types


I added my unit tests but since AECU doesn't use io.wcm or junit5-jupiter, I commented them all out.  You can still see how it is tested.  I just couldn't bring myself to recreating the tests.
